_posts/2024-01-25-websockets-in-http4s.md

```diff
@@ -0,0 +1,1561 @@
+
+If we run our application at this point and try to send messages, they will be received from the server and appended to the output `<div>` proving that our WebSocket server and client are connected and working.
+
+## 5. Chat Application
+In this section, we build a chat application using http4s WebSocket implementation, the code in the following sections is inspired by [work](https://github.com/MartinSnyder/http4s-chatserver) done by [Martin Snyder](https://github.com/MartinSnyder). 
+
+### 5.1 Topic for message pipeline
+
+If we open two pages in our browser and navigate to `chat.html`, when we try to send messages in one, they won't appear in the other, in other words, each user is chatting with his/herself.
+
+To fix this issue we'll need a `Topic` which is a concurrency primitive from Fs2 used to implement a publish-subscribe patterns:
+
+```scala
+object Routes {
+  def service[F[_]: Files: MonadThrow: Concurrent](
+      wsb: WebSocketBuilder2[F],
+      q: Queue[F, WebSocketFrame],
+      t: Topic[F, WebSocketFrame]
+  ): HttpApp[F] = {
+    val dsl = new Http4sDsl[F] {}
+    import dsl.*
+    HttpRoutes.of[F] {
+      ...
+      case GET -> Root / "ws" =>
+        val send: Stream[F, WebSocketFrame] =
+          t.subscribe(maxQueued = 1000)
+
+        val receive: Pipe[F, WebSocketFrame, Unit] =
+          _.foreach(q.offer)
+
+        wsb.build(send, receive)
+    }
+  }.orNotFound
+}
+```
+
+Here we promote, the `Queue` and `Topic` to the top level by passing them as parameters to the `service()` function, the `Queue` is still needed to order messages as First In First Out (FIFO). 
+
+Furthermore, the `receive` handle remains unchanged but the `send` handle now subscribes to the `Topic` using the `subscribe()` method which takes `maxQueued` as a parameter. This is the maximum number of elements to enqueue to the subscription queue before blocking publishers.
+
+We'll also need to update `Server.scala` and pass Topic and Queue as parameters:
+
+```diff
++ import fs2.concurrent.Topic
++ import org.http4s.websocket.WebSocketFrame
++ import cats.effect.std.Queue
+
+  object Server{
+    def server[F[_]: Async: Files: Network](
+      q: Queue[F, WebSocketFrame],
+      t: Topic[F, WebSocketFrame]
+    ): F[Unit] =
+      val host = host"0.0.0.0"
+      val port = port"8080"
+      EmberServerBuilder
+        .default[F]
+        .withHost(host)
+        .withPort(port)
+-       .withHttpWebSocketApp(wsb => service(wsb))
++       .withHttpWebSocketApp(wsb => service(wsb, q, t))
+        .build
+        .useForever
+        .void
+  }
+```
+
+We do the same to `Program.scala` with some extra updates:
+
+```scala
+import fs2.concurrent.Topic
+import org.http4s.websocket.WebSocketFrame
+import cats.effect.std.Queue
+import fs2.Stream
+
+object Program extends IOApp.Simple{
+  def program: IO[Unit] =
+    for
```

Comment:
Let's use braces instead of the indentation syntax, please. Several people have reported issues with indentation with previous blog posts.

Reply:
Fixed